### ISSUE TITLE
ForexFactory study causing CPU exception

### ISSUE THREAD

**Issue by alfakid** [gcUserStudies/sierrachart#1]
Hello,

Love the studies! Thanks so much for sharing. I have recently started experiencing this issue...

2024-08-19  18:53:35.767 | Warning: The Custom DLL study "gcUserStudies.scsf_gcForexFactory" has just caused a CPU exception.

Current env details:
SC: 2667
OpenGL enabled

Previous:
SC: 2652

Upgrade does not appear to have resolved the issue. I will try updating NVIDIA driver next.

**Comment by alfakid** [comment]
After further review, looks like maybe this study is crashing because Forex Factory is throttling the requests?


					<ul class="error__body">
						<li>
							You've exceeded the limit for Calendar Export requests. Please wait five minutes before trying again.
				


							The Calendar export file is only updated once per hour.  Requesting it more than that is unnecessary and can result in being blocked.

**Comment by svltvn** [comment]
+1

**Comment by alfakid** [comment]
I set my study settings to only check for updates from FF every 60 minutes, but it doesn't seem that the study honors this setting. I don't use time based charts, perhaps that is part of the problem?

**Comment by svltvn** [comment]
I tried a couple things here, and was able to get it working:
- Remove study from all charts it's attached to, and then re-add it to only one chart
- Set the Check for Update Interval to 60m (Don't know if this matters too much, but I am using this on a non-time based chart as well)
- Hit Apply, but the error may pop up again, so I waited a few minutes and then hit apply on it again (I don't remove the study, I just opened up the study window again and hit apply once)

I think you are correct in the fact that FF is throttling the amount of calls that can be made to the website, as to why I went with just one chart only and waited before applying the study again. Anytime you hit apply, it will send a request to the website

**Comment by gcUserStudies** [comment]
Here is a pre-release version that should resolve the Forex Factory issues. Can you please test this version out. 

https://github.com/gcUserStudies/tmp/raw/main/gcUserStudies_rc4.zip

**Comment by svltvn** [comment]
My good sir! Thank you for the update, was not expected! Currently testing it and working as expected on multiple charts, and did have to remove the old dll and replace with new one. Thanks again and have been meaning to send you a thumbs up on your studies in general, keep up the great work! Lmk if you need me to do any more additional testing
![{DC1827FC-8807-4A75-AB89-F20AF464CD8A}](https://github.com/user-attachments/assets/c028d74c-6c25-4f4d-828b-b30dd68af62f)
![{64E19E98-1B7E-400B-BC14-52BCA078D11A}](https://github.com/user-attachments/assets/00203675-d3fb-4dc8-a508-43e0c7afd2b2)

**Comment by svltvn** [comment]
Also you sent over some other studies in that DLL above. These other studies look really great as well, was wondering if you would be also releasing the source code? Also any tips on how to improve with ASCIL, your studies are absolutely brilliant and hope to learn and give back to the community like you do!

**Comment by alfakid** [comment]
So far so good for me as well! My only suggestion, and beggars can't be choosers, would be to allow for the json file location to be specified so that multiple sc instances can use the same file. Maybe one instance of the study could be set to fetch from FF and the rest (operating in other charts in other instances of SC) just get to read the cached json. Maybe some day when you are bored ;)

Love your work! I appreciate you!

**Comment by gcUserStudies** [comment]
> My good sir! Thank you for the update, was not expected! Currently testing it and working as expected on multiple charts, and did have to remove the old dll and replace with new one. Thanks again and have been meaning to send you a thumbs up on your studies in general, keep up the great work! Lmk if you need me to do any more additional testing

Yes. I have been working on splitting out the main file into multiple study files, includes, and all of that so it's easier to deal with. As soon as I'm finished I will post it all

**Comment by gcUserStudies** [comment]
> So far so good for me as well! My only suggestion, and beggars can't be choosers, would be to allow for the json file location to be specified so that multiple sc instances can use the same file. Maybe one instance of the study could be set to fetch from FF and the rest (operating in other charts in other instances of SC) just get to read the cached json. Maybe some day when you are bored ;)
> 
> Love your work! I appreciate you!

This is a great idea. For some reason I was thinking Sierra only had access to the Data folder for reading/writing files. Upon further looking that doesn't seem to be the case. I will take a look at this and do some testing on it.
